engine/src/main/java/org/pentaho/di/trans/steps/formula/RowForumulaContext.java

```diff
@@ -0,0 +1,212 @@
+import org.pentaho.reporting.libraries.formula.ErrorValue;
+import org.pentaho.reporting.libraries.formula.EvaluationException;
+import org.pentaho.reporting.libraries.formula.FormulaContext;
+import org.pentaho.reporting.libraries.formula.LibFormulaErrorValue;
+import org.pentaho.reporting.libraries.formula.LocalizationContext;
+import org.pentaho.reporting.libraries.formula.function.FunctionRegistry;
+import org.pentaho.reporting.libraries.formula.operators.OperatorFactory;
+import org.pentaho.reporting.libraries.formula.typing.Type;
+import org.pentaho.reporting.libraries.formula.typing.TypeRegistry;
+import org.pentaho.reporting.libraries.formula.typing.coretypes.AnyType;
+import org.pentaho.reporting.libraries.formula.typing.coretypes.NumberType;
+import org.pentaho.reporting.libraries.formula.typing.coretypes.TextType;
+
+public class RowForumulaContext implements FormulaContext {
+  private RowMetaInterface rowMeta;
+  private FormulaContext formulaContext;
+  private Map<String, Integer> valueIndexMap;
+  private Object[] rowData;
+
+  public RowForumulaContext( RowMetaInterface row ) {
+    this.formulaContext = new DefaultFormulaContext();
+    this.rowMeta = row;
+    this.rowData = null;
+    this.valueIndexMap = new Hashtable<>();
+  }
+
+  public Type resolveReferenceType( Object name ) {
+    if ( name instanceof String ) {
+      ValueMetaInterface valueMeta = this.rowMeta.searchValueMeta( (String) name );
+      if ( valueMeta != null ) {
+        switch ( valueMeta.getType() ) {
+          case ValueMetaInterface.TYPE_STRING:
+            return TextType.TYPE;
+          case ValueMetaInterface.TYPE_INTEGER:
+          case ValueMetaInterface.TYPE_BIGNUMBER:
+          case ValueMetaInterface.TYPE_NUMBER:
+            return NumberType.GENERIC_NUMBER;
+          default:
+            return AnyType.TYPE;
+        }
+      }
+    }
+    return AnyType.TYPE;
+  }
+
+  /**
+   * We return the content of a Value with the given name. We cache the position of the field indexes.
+   *
+   * @see org.jfree.formula.FormulaContext#resolveReference(java.lang.Object)
+   */
+  public Object resolveReference( Object name ) throws EvaluationException {
+    if ( name instanceof String ) {
+      ValueMetaInterface valueMeta;
+      Integer idx = valueIndexMap.get( name );
+      if ( idx != null ) {
+        valueMeta = rowMeta.getValueMeta( idx.intValue() );
+      } else {
+        int index = rowMeta.indexOfValue( (String) name );
+        if ( index < 0 ) {
+          ErrorValue errorValue = new LibFormulaErrorValue( LibFormulaErrorValue.ERROR_INVALID_ARGUMENT );
+          throw new EvaluationException( errorValue );
+        }
+        valueMeta = rowMeta.getValueMeta( index );
+        idx = new Integer( index );
+        valueIndexMap.put( (String) name, idx );
+      }
+      Object valueData = rowData[idx];
+      try {
+        return getPrimitive( valueMeta, valueData );
+      } catch ( KettleValueException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)